internal/pkg/service/common/etcdclient/etcdclient.go

```diff
@@ -0,0 +1,228 @@
+		c.waitGroup = v
+	}
+}
+
+// New creates new etcd client.
+// The client terminates the connection when the context is done.
+func New(ctx context.Context, tracer trace.Tracer, endpoint, namespace string, opts ...Option) (c *etcd.Client, err error) {
+	ctx, span := tracer.Start(ctx, "kac.api.server.templates.dependencies.EtcdClient")
+	defer telemetry.EndSpan(span, &err)
+
+	// Apply options
+	conf := config{
+		endpoint:          endpoint,
+		namespace:         namespace,
+		connectCtx:        ctx,
+		connectTimeout:    defaultConnectionTimeout,
+		keepAliveTimeout:  defaultKeepAliveTimeout,
+		keepAliveInterval: defaultKeepAliveInterval,
+		autoSyncInterval:  defaultAutoSyncInterval,
+	}
+	for _, o := range opts {
+		o(&conf)
+	}
+
+	// Trim and validate
+	endpoint = strings.Trim(endpoint, " /")
+	if endpoint == "" {
+		return nil, errors.New("etcd endpoint is not set")
+	}
+	namespace = strings.Trim(namespace, " /") + "/"
+	if namespace == "/" {
+		return nil, errors.New("etcd namespace is not set")
+	}
+
+	// Setup logger
+	logger := conf.logger
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	// Create a zap logger for etcd client
+	encoder := zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig())
+	etcdLogger := zap.New(log.NewCallbackCore(func(entry zapcore.Entry, fields []zapcore.Field) {
+		// Add component=etcd-client field
+		fields = append(fields, zapcore.Field{Key: "component", String: "etcd-client", Type: zapcore.StringType})
+
+		// Encode and log message
+		if bytes, err := encoder.EncodeEntry(entry, fields); err == nil {
+			logger.Log(entry.Level.String(), strings.TrimRight(bytes.String(), "\n"))
+		} else {
+			logger.Warnf("cannot log msg from etcd client: %s", err)
+		}
+	}))
+
+	// Create connect context
+	connectCtx, connectCancel := context.WithTimeout(conf.connectCtx, conf.connectTimeout)
+	defer connectCancel()
+
+	// Create client
+	startTime := time.Now()
+	logger.Infof("connecting to etcd, connectTimeout=%s, keepAliveTimeout=%s, keepAliveInterval=%s", conf.connectTimeout, conf.keepAliveTimeout, conf.keepAliveInterval)
+	c, err = etcd.New(etcd.Config{
+		Context:              ctx, // !!! a long-lived context must be used, client exists as long as the entire server
+		Endpoints:            []string{endpoint},
+		AutoSyncInterval:     conf.autoSyncInterval,
+		DialTimeout:          conf.connectTimeout,
+		DialKeepAliveTimeout: conf.keepAliveTimeout,
+		DialKeepAliveTime:    conf.keepAliveInterval,
+		Username:             conf.username, // optional
+		Password:             conf.password, // optional
+		Logger:               etcdLogger,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // wait for the connection
+			grpc.WithReturnConnectionError(),
+			grpc.WithConnectParams(grpc.ConnectParams{
+				Backoff: backoff.Config{
+					BaseDelay:  100 * time.Millisecond,
+					Multiplier: 1.5,
+					Jitter:     0.2,
+					MaxDelay:   15 * time.Second,
+				},
+			}),
+		},
+	})
```

Comment:
Manualne som to skusal a upravil som default retries, ktore cakali max `120s`, co je prilis vela.

Reply:
timeout aj znovu-nadviazanie spojenia funguje dobre:
![image](https://user-images.githubusercontent.com/19371734/203250937-f5591f3e-148c-4815-9afc-b80d015dd118.png)